ledger/src/entry.rs

```diff
@@ -148,53 +148,71 @@ pub fn next_hash(start_hash: &Hash, num_hashes: u64, transactions: &[Transaction
     }
 }
 
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum EntryVerificationStatus {
+    Pending,
+    Success,
+    Failure,
+}
+
 pub struct EntryVerifyState {
     thread_h: Option<JoinHandle<u64>>,
     hashes: Option<Arc<Mutex<Vec<Hash>>>>,
-    verified: bool,
+    verification_status: EntryVerificationStatus,
     tx_hashes: Vec<Option<Hash>>,
     start_time_ms: u64,
 }
 
 impl EntryVerifyState {
-    pub fn finish_verify(&mut self, entries: &[Entry]) -> bool {
-        if self.hashes.is_some() {
-            let gpu_time_ms = self.thread_h.take().unwrap().join().unwrap();
-
-            let mut verify_check_time = Measure::start("verify_check");
-            let hashes = self.hashes.take().expect("hashes.as_ref");
-            let hashes = Arc::try_unwrap(hashes)
-                .expect("unwrap Arc")
-                .into_inner()
-                .expect("into_inner");
-            let res = PAR_THREAD_POOL.with(|thread_pool| {
-                thread_pool.borrow().install(|| {
-                    hashes
-                        .into_par_iter()
-                        .zip(&self.tx_hashes)
-                        .zip(entries)
-                        .all(|((hash, tx_hash), answer)| {
-                            if answer.num_hashes == 0 {
-                                hash == answer.hash
-                            } else {
-                                let mut poh = Poh::new(hash, None);
-                                if let Some(mixin) = tx_hash {
-                                    poh.record(*mixin).unwrap().hash == answer.hash
+    pub fn status(&self) -> EntryVerificationStatus {
+        self.verification_status
+    }
+
+    pub fn finish_verify(&mut self, entries: &[Entry]) -> EntryVerificationStatus {
+        match self.verification_status {
```

Comment:
@ParthDesai can we just use `if verified` here, instead of defining a new `enum`?

Reply:
@pgarg66  Simply using boolean and mark `verified=false`, there will be ambiguity among these two:
1) Pending verification
2) Failed verification

Advantage of using enum is we can represent all three state without ambiguity:
Pending/InProgress (In case of gpu)
Success
Failure

Reply:
`verified` would be false only if `verify_cpu` (or GPU) failed in verification. On that note, https://github.com/solana-labs/solana/pull/7169/files#diff-376f67abc475fe7f487984f195d0fca7L327 could set verified to `res` (similar to `verify_cpu`), instead of `false`.
 
@sakridge , do you know why `verified: false` at  https://github.com/solana-labs/solana/pull/7169/files#diff-376f67abc475fe7f487984f195d0fca7L327. `finish_verify` would return false if `self.hashes` is `None`. Is that the intended behavior?

Reply:
@pgarg66 How can `self.hashes` be `None` in that case, it sets it to `Some(hashes)`. What it sets it to should actually never matter since it should go through the GPU path there and using this:
https://github.com/solana-labs/solana/blob/323673c3c0b8e33b14d27f1e408a4624c3eb5160/ledger/src/entry.rs#L179

Reply:
@ParthDesai we could keep the return here as bool, no? We should never be pending after `finish_verify`

Reply:
> @pgarg66 How can `self.hashes` be `None` in that case, it sets it to `Some(hashes)`. What it sets it to should actually never matter since it should go through the GPU path there and using this:
> https://github.com/solana-labs/solana/blob/323673c3c0b8e33b14d27f1e408a4624c3eb5160/ledger/src/entry.rs#L179

Maybe a better question would've been: would setting `verified: res` in  https://github.com/solana-labs/solana/pull/7169/files#diff-376f67abc475fe7f487984f195d0fca7L327 have any adverse effect? I guess not, based on your reply above.

Reply:
@sakridge @pgarg66  I did it for consistency's sake. 

My reasoning comes because of different handling of object states:

In case of CPU verification:
Verification happens in `start_verify` itself, and verified is set to `true` or `false`

In case of GPU verification:
Verification happens when `finish_verify` is called and then verified is changed to `true` in case verification succeeds, and set to `false` in case of verification fails. 

So, the semantics of `verified: false` after `start_verify` is called depends upon whether it is CPU or GPU verification.

Maybe, instead of checking if `Hashes` is `None` or `Some(hashes)` to determine the behaviour we could just have a flag indicating CPU or GPU verification.

Reply:
@pgarg66 there is no `res` there. At that point we have just kicked off the GPU path and we don't know if it will be verified until finish_verify is called.

Reply:
Got it.. makes sense. Thanks

Reply:
@sakridge So, if I use `bool`, how would I be able to distinguish between verification completed or pending? after my call to `start_verify` returns?

Reply:
@ParthDesai For `start_verify` you can use the enum since the state can be in pending state there. But for when you know you are done, I think it makes it cleaner from the API to only expose two options passed or not.

Reply:
I think it might be also cleaner to use an enum, and then have a CPU struct and a GPU struct contained in them so it's clear which path is using what fields.

Reply:
@pgarg66 @sakridge How does it look like now? Main interface is left as it is, while also allowing different enum for CPU and GPU verification.